### PR TITLE
Yfi adapter

### PR DESCRIPTION
**Pull request by Yorkemartin**

### CODE REVIEW

**Comment by Yorkemartin** [comment]
need to fix rate calculation in the getComponents function. currently need to pass address of account to get the rate calculated and returned, but unsure how to do it

**Comment by Yorkemartin** [comment]
There is no balanceOf that is readable for staked YFI at pools, but
there is a balanceOf for the rewards earned, which can be queried. 

YFI adapter reads .earned(account) of each of the three currently available YFI staking contracts

todo - update test address to reflect three different instances of YFI 
staking address holders

todo - update test to return correct balances of staking reward 
addresses - if each balance is

File: contracts/adapters/yearn/YFIAdapter.sol [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+pragma solidity 0.6.5;
+pragma experimental ABIEncoderV2;
+
+import { ERC20 } from "../../ERC20.sol";
+import { ProtocolAdapter } from "../ProtocolAdapter.sol";
+
+
+
+
```
Comment (sobolev-igor) [inline contracts/adapters/yearn/YFIAdapter.sol:24]: leave only 2 lines

File: contracts/adapters/yearn/YFIAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+ * Only the functions required for YFIAdapter contract are added.
+ * The YearnRewards Curve Pool contract is available here
+ * 0xcc9EFea3ac5Df6AD6A656235Ef955fBfEF65B862
+ * The YearnRewards Balancer Pool contract is available here
+ * 0x033E52f513F9B98e129381c6708F9faA2DEE5db5
+ */
+
+interface YearnRewards {
+    function earned(address) external view returns (uint256);
+}
```
Comment (sobolev-igor) [inline contracts/adapters/yearn/YFIAdapter.sol:51]:
```suggestion
/**
 * @dev StakingRewards contract interface.
 * Only the functions required for YearnStakingV1Adapter contract are added.
 * The StakingRewards contract is available here
 * github.com/Synthetixio/synthetix/blob/master/contracts/StakingRewards.sol.
 */
interface StakingRewards {
    function earned(address) external view returns (uint256);
}
```

File: contracts/adapters/yearn/YFIAdapter.sol [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+ * Only the functions required for YFIAdapter contract are added.
+ * The YearnRewards contract is available here
+ * 0xcc9EFea3ac5Df6AD6A656235Ef955fBfEF65B862
+ */
+
+/*
+interface YearnRewards {
+    function claimable(address) external view returns (uint);
+}
+*/
```
Comment (sobolev-igor) [inline contracts/adapters/yearn/YFIAdapter.sol:38]: One interface is enough

File: contracts/adapters/yearn/YFIAdapter.sol (hunk 2 above)
Comment (sobolev-igor) [reply contracts/adapters/yearn/YFIAdapter.sol:51]: this interface is enough as they use Synthetix's contracts

File: contracts/adapters/yearn/YFIAdapter.sol [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+ * @dev YearnGovernance contract interface.
+ * Only the functions required for YFIAdapter contract are added.
+ * The YearnRewards Governance contract is available here
+ * 0x3A22dF48d84957F907e67F4313E3D43179040d6E
+ */
+
+interface YearnGovernance {
+    function earned(address) external view returns (uint256);
+}
+
```
Comment (sobolev-igor) [inline contracts/adapters/yearn/YFIAdapter.sol:63]: remove

File: contracts/adapters/yearn/YFIAdapter.sol [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+ * @title Adapter for Yearn.finance YFI Staking protocol.
+ * @dev Implementation of ProtocolAdapter interface.
+ * @author Connor Martin <cnr.mrtn@gmail.com>
+ */
+
+contract YFIAdapter is ProtocolAdapter {
+
+    string public constant override adapterType = "Asset";
+
+    string public constant override tokenType = "YFIToken";
```
Comment (sobolev-igor) [inline contracts/adapters/yearn/YFIAdapter.sol:76]:
```suggestion
    string public constant override tokenType = "ERC20";
```

File: contracts/adapters/yearn/YFIAdapter.sol [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+
+
+/**
+ * @title Adapter for Yearn.finance YFI Staking protocol.
+ * @dev Implementation of ProtocolAdapter interface.
+ * @author Connor Martin <cnr.mrtn@gmail.com>
+ */
+
+contract YFIAdapter is ProtocolAdapter {
```
Comment (sobolev-igor) [inline contracts/adapters/yearn/YFIAdapter.sol:72]:
```suggestion
contract YearnStakingV1Adapter is ProtocolAdapter {
```

File: contracts/adapters/yearn/YFIAdapter.sol [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+    string public constant override adapterType = "Asset";
+
+    string public constant override tokenType = "YFIToken";
+
+    address internal constant YEARNREWARDS_STAKING_CURVEPOOL = 0x0001FB050Fe7312791bF6475b96569D83F695C9f;
+    address internal constant YEARNREWARDS_STAKING_BALANCERPOOL = 0x033E52f513F9B98e129381c6708F9faA2DEE5db5;
+    address internal constant YEARNREWARDS_STAKING_GOVERNANCEPOOL = 0x3A22dF48d84957F907e67F4313E3D43179040d6E;
+    //address internal constant YEARNREWARDS_CONTRACT = 0xcc9EFea3ac5Df6AD6A656235Ef955fBfEF65B862;
+    address internal constant YFI = 0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e;
```
Comment (sobolev-igor) [inline contracts/adapters/yearn/YFIAdapter.sol:82]:
```suggestion
    address internal constant YFI = 0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e;
    address internal constant CURVE_Y = 0xdF5e0e81Dff6FAF3A7e52BA697820c5e32D806A8;
    address internal constant BALANCER_DAI_YFI_98_2 = 0x60626db611a9957C1ae4Ac5b7eDE69e24A3B76c5;
    address internal constant BALANCER_YFI_CURVE_Y_2_98 = 0x95C4B6C7CfF608c0CA048df8b81a484aA377172B;
    
    address internal constant CURVE = 0x0001FB050Fe7312791bF6475b96569D83F695C9f;
    address internal constant BALANCER = 0x033E52f513F9B98e129381c6708F9faA2DEE5db5;
    address internal constant GOVERNANCE = 0x3A22dF48d84957F907e67F4313E3D43179040d6E;
    address internal constant FEE_REWARDS = 0xb01419E74D8a2abb1bbAD82925b19c36C191A701;
```

Comment (sobolev-igor) [reply contracts/adapters/yearn/YFIAdapter.sol:82]: just changed names to comply with their website

File: contracts/adapters/yearn/YFIAdapter.sol [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+          return YearnRewards(YEARNREWARDS_CONTRACT).claimable(account);
+      }  else */if (token == YEARNREWARDS_STAKING_CURVEPOOL) {
+          return YearnRewards(YEARNREWARDS_STAKING_CURVEPOOL).earned(account);
+      } else if (token == YEARNREWARDS_STAKING_BALANCERPOOL) {
+          return YearnRewards(YEARNREWARDS_STAKING_BALANCERPOOL).earned(account);
+      }  else if (token == YEARNREWARDS_STAKING_GOVERNANCEPOOL) {
+          return YearnGovernance(YEARNREWARDS_STAKING_GOVERNANCEPOOL).earned(account);
+      }  else   {
+        return 0;
+      }
```
Comment (sobolev-igor) [inline contracts/adapters/yearn/YFIAdapter.sol:101]:
```suggestion
      if (token == YFI) {
          uint256 totalRewards = 0;
          totalRewards += ERC20(CURVE).earned(account);
          totalRewards += ERC20(BALANCER).earned(account);
          totalRewards += ERC20(GOVERNANCE).earned(account);
          totalRewards += ERC20(FEE_REWARDS).balanceOf(account);
          return totalRewards;
      } else if (token == CURVE_Y) {
          return ERC20(CURVE).balanceOf(account);
          return StakingRewards(FEE_REWARDS).earned(account);
      } else if (token == BALANCER_DAI_YFI_98_2) {
          return ERC20(BALANCER).balanceOf(account);
      } else if (token == BALANCER_YFI_CURVE_Y_2_98) {
          return ERC20(GOVERNANCE).balanceOf(account);
      } else {
          return 0;
      }
```

File: migrations_scripts/1_deploy_registry_and_add_adapters.js [hunk 9]
```diff
@@ -675,6 +688,18 @@ module.exports = async (deployer, network, accounts) => {
     '0',
   ]);
 
+  await deployer.deploy(YFIAdapter, { from: accounts[0] });
+  adapters.push([YFIAdapter.address]);
+  tokens.push([yfiAdapterTokens]);
+  protocolNames.push('YFI Governance');
```
Comment (sobolev-igor) [inline migrations_scripts/1_deploy_registry_and_add_adapters.js:694]:
```suggestion
  protocolNames.push('Yearn Staking V1');
```

File: migrations_scripts/1_deploy_registry_and_add_adapters.js [hunk 10]
```diff
@@ -155,6 +155,12 @@ const saiPoolAddress = '0xb7896fce748396EcFC240F5a0d3Cc92ca42D7d84';
 const daiPoolAddress = '0x29fe7D60DdF151E5b52e5FAB4f1325da6b2bD958';
 const usdcPoolAddress = '0x0034Ea9808E620A0EF79261c51AF20614B742B24';
 
+const yfiAddress = '0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e';
+const yearnRewardsStakingCurvePoolAddress = '0x0001FB050Fe7312791bF6475b96569D83F695C9f';
+const yearnRewardsStakingBalancerPoolAddress = '0x033E52f513F9B98e129381c6708F9faA2DEE5db5';
+const yearnRewardsStakingGovernancePoolAddress = '0x3A22dF48d84957F907e67F4313E3D43179040d6E';
+const yearnRewardsContractAddress = '0xcc9EFea3ac5Df6AD6A656235Ef955fBfEF65B862';
```
Comment (sobolev-igor) [inline migrations_scripts/1_deploy_registry_and_add_adapters.js:162]: fix to comply with contracts

YFI, CURVE_Y, BALANCER_DAI_YFI_98_2, BALANCER_YFI_CURVE_Y_2_98

File: migrations_scripts/1_deploy_registry_and_add_adapters.js [hunk 11]
```diff
@@ -333,6 +339,13 @@ const synthetixAssetAdapterTokens = [
 const synthetixDebtAdapterTokens = [
   susdAddress,
 ];
+const yfiAdapterTokens = [
+  yfiAddress,
+  yearnRewardsStakingCurvePool,
+  yearnRewardsStakingBalancerPool,
+  yearnRewardsStakingGovernancePool,
+  yearnRewardsContract,
```
Comment (sobolev-igor) [inline migrations_scripts/1_deploy_registry_and_add_adapters.js:347]: fix

File: migrations_scripts/1_deploy_registry_and_add_adapters.js [hunk 12]
```diff
@@ -819,6 +844,7 @@ module.exports = async (deployer, network, accounts) => {
           'SmartToken',
           'Uniswap V1 pool token',
           'Uniswap V2 pool token',
+          'YFI Token',
```
Comment (sobolev-igor) [inline migrations_scripts/1_deploy_registry_and_add_adapters.js:847]: no need to add token adapter

File: contracts/adapters/yearn/YFIAdapter.sol (hunk 2 above)
Comment (Yorkemartin) [reply contracts/adapters/yearn/YFIAdapter.sol:51]: Thanks - I thought I had to use the specific interface/contract name of the contract pointed to in the 'address internal constant' declaration

File: migrations_scripts/1_deploy_registry_and_add_adapters.js (hunk 10 above)
Comment (Yorkemartin) [reply migrations_scripts/1_deploy_registry_and_add_adapters.js:162]: fixed but unsure of if I should capitalize the first word to reflect the contract declaration or lowercase to reflect the other addresses in the migrations file

File: contracts/adapters/yearn/YFIAdapter.sol [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    address internal constant BALANCER = 0x033E52f513F9B98e129381c6708F9faA2DEE5db5;
+    address internal constant GOVERNANCE = 0x3A22dF48d84957F907e67F4313E3D43179040d6E;
+    address internal constant FEE_REWARDS = 0xb01419E74D8a2abb1bbAD82925b19c36C191A701;
+
+    /**
+     * @return Amount of YFI rewards earned after staking in a given account.
+     * @dev Implementation of ProtocolAdapter interface function.
+     */
+
+
```
Comment (sobolev-igor) [inline contracts/adapters/yearn/YFIAdapter.sol:61]: remove

File: migrations_scripts/1_deploy_registry_and_add_adapters.js [hunk 14]
```diff
@@ -155,6 +155,11 @@ const saiPoolAddress = '0xb7896fce748396EcFC240F5a0d3Cc92ca42D7d84';
 const daiPoolAddress = '0x29fe7D60DdF151E5b52e5FAB4f1325da6b2bD958';
 const usdcPoolAddress = '0x0034Ea9808E620A0EF79261c51AF20614B742B24';
 
+const yfi = '0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e';
+const curve_Y = '0xdF5e0e81Dff6FAF3A7e52BA697820c5e32D806A8';
+const balancer_DAI_YFI_98_2 = '0x60626db611a9957C1ae4Ac5b7eDE69e24A3B76c5';
+const balancer_YFI_CURVE_Y_2_98 = '0x95C4B6C7CfF608c0CA048df8b81a484aA377172B';
+
```
Comment (sobolev-igor) [inline migrations_scripts/1_deploy_registry_and_add_adapters.js:162]:
```suggestion
const yfi = '0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e';
const balancerDai98Yfi2 = '0x60626db611a9957C1ae4Ac5b7eDE69e24A3B76c5';
const balancerYfi2yCrv98 = '0x95C4B6C7CfF608c0CA048df8b81a484aA377172B';

```

Comment (sobolev-igor) [reply migrations_scripts/1_deploy_registry_and_add_adapters.js:162]: yCrv is already in the file (line 141)

File: migrations_scripts/1_deploy_registry_and_add_adapters.js [hunk 15]
```diff
@@ -333,6 +338,12 @@ const synthetixAssetAdapterTokens = [
 const synthetixDebtAdapterTokens = [
   susdAddress,
 ];
+const yfiAdapterTokens = [
+  yfi,
+  curve_Y,
+  balancer_DAI_YFI_98_2,
+  balancer_YFI_CURVE_Y_2_98,
```
Comment (sobolev-igor) [inline migrations_scripts/1_deploy_registry_and_add_adapters.js:345]:
```suggestion
  yCrv,
  balancerDai98Yfi2,
  balancerYfi2yCrv98,
```

File: migrations_scripts/1_use_registry_and_add_adapters.js [hunk 16]
```diff
@@ -155,6 +155,11 @@ const saiPoolAddress = '0xb7896fce748396EcFC240F5a0d3Cc92ca42D7d84';
 const daiPoolAddress = '0x29fe7D60DdF151E5b52e5FAB4f1325da6b2bD958';
 const usdcPoolAddress = '0x0034Ea9808E620A0EF79261c51AF20614B742B24';
 
+const yfi = '0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e';
+const curve_Y = '0xdF5e0e81Dff6FAF3A7e52BA697820c5e32D806A8';
+const balancer_DAI_YFI_98_2 = '0x60626db611a9957C1ae4Ac5b7eDE69e24A3B76c5';
+const balancer_YFI_CURVE_Y_2_98 = '0x95C4B6C7CfF608c0CA048df8b81a484aA377172B';
+
```
Comment (sobolev-igor) [inline migrations_scripts/1_use_registry_and_add_adapters.js:162]:
```suggestion
const yfi = '0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e';
const balancerDai98Yfi2 = '0x60626db611a9957C1ae4Ac5b7eDE69e24A3B76c5';
const balancerYfi2yCrv98 = '0x95C4B6C7CfF608c0CA048df8b81a484aA377172B';

```

File: migrations_scripts/1_use_registry_and_add_adapters.js [hunk 17]
```diff
@@ -333,6 +338,12 @@ const synthetixAssetAdapterTokens = [
 const synthetixDebtAdapterTokens = [
   susdAddress,
 ];
+const yfiAdapterTokens = [
+  yfi,
+  curve_Y,
+  balancer_DAI_YFI_98_2,
+  balancer_YFI_CURVE_Y_2_98,
```
Comment (sobolev-igor) [inline migrations_scripts/1_use_registry_and_add_adapters.js:345]:
```suggestion
  yCrv,
  balancerDai98Yfi2,
  balancerYfi2yCrv98,
```

File: package.json [hunk 18]
```diff
@@ -37,7 +38,7 @@
     "eslint-plugin-import": "2.20.2",
     "eth-gas-reporter": "0.2.17",
     "ethlint": "1.2.5",
-    "ganache-cli": "6.9.1",
+    "ganache-cli": "^6.9.1",
```
Comment (sobolev-igor) [inline package.json:41]: rollback changes

File: package.json [hunk 19]
```diff
@@ -27,7 +27,8 @@
     "core-js": "3.6.5",
     "dotenv": "8.2.0",
     "regenerator-runtime": "0.13.5",
-    "truffle": "5.1.31",
+    "remixd": "^0.2.3-alpha.4",
+    "truffle": "^5.1.31",
```
Comment (sobolev-igor) [inline package.json:31]: rollback changes

File: test/YFIAdapter.js [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+const ProtocolAdapter = artifacts.require('YFIAdapter');
+const TokenAdapter = artifacts.require('TokenSetsTokenAdapter');
+const ERC20TokenAdapter = artifacts.require('ERC20TokenAdapter');
+
+
+contract.only('YFIAdapter', () => {
+  const yfi = '0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e';
+  const curve_Y = '0xdF5e0e81Dff6FAF3A7e52BA697820c5e32D806A8';
+  const balancer_DAI_YFI_98_2 = '0x60626db611a9957C1ae4Ac5b7eDE69e24A3B76c5';
+  const balancer_YFI_CURVE_Y_2_98 = '0x95C4B6C7CfF608c0CA048df8b81a484aA377172B';
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:13]:
```suggestion
  const yCrvAddress = '0xdF5e0e81Dff6FAF3A7e52BA697820c5e32D806A8';
  const balancerDai98Yfi2Address = '0x60626db611a9957C1ae4Ac5b7eDE69e24A3B76c5';
  const balancerYfi2yCrv98Address = '0x95C4B6C7CfF608c0CA048df8b81a484aA377172B';
```

File: test/YFIAdapter.js [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+
+  let accounts;
+  let adapterRegistry;
+  let protocolAdapterAddress;
+  let tokenAdapterAddress;
+  let erc20TokenAdapterAddress;
+  const yfi = [
+    yfi,
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:21]:
```suggestion
    yfiAddress,
```

File: test/YFIAdapter.js [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+    'yearn.finance',
+    'YFI',
+    '18',
+  ];
+  const curve_Y = [
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:26]:
```suggestion
  const yCrv = [
```

File: test/YFIAdapter.js [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+    curve_Y,
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:27]:
```suggestion
    yCrvAddress,
```

File: test/YFIAdapter.js [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+    'Not available',
+    'N/A',
+    '0',
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:30]: should not be "N/A", ERC20TokenAdapter should fetch it correctly

File: test/YFIAdapter.js [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+  ];
+  const balancer_DAI_YFI_98_2 = [
+    balancer_DAI_YFI_98_2,
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:33]:
```suggestion
  const balancerDai98Yfi2 = [
    balancerDai98Yfi2Address,
```

File: test/YFIAdapter.js [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+    'Not available',
+    'N/A',
+    '0',
+  ];
+  const balancer_YFI_CURVE_Y_2_98 = [
+    balancer_YFI_CURVE_Y_2_98,
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:39]:
```suggestion
  const balancerYfi2yCrv98 = [
    balancerYfi2yCrv98Address,
```

File: test/YFIAdapter.js [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+  beforeEach(async () => {
+    accounts = await web3.eth.getAccounts();
+    await ProtocolAdapter.new({ from: accounts[0] })
+      .then((result) => {
+        protocolAdapterAddress = result.address;
+      });
+    await TokenAdapter.new({ from: accounts[0] })
+      .then((result) => {
+        tokenAdapterAddress = result.address;
+      });
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:54]: remove

File: test/YFIAdapter.js [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+        '0',
+      ]],
+      [[
+        protocolAdapterAddress,
+      ]],
+      [[[
+        yfi,
+        curve_Y,
+        balancer_DAI_YFI_98_2,
+        balancer_YFI_CURVE_Y_2_98,
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:79]:
```suggestion
        yCrv,
        balancerDai98Yfi2,
        balancerYfi2yCrv98,
```

File: test/YFIAdapter.js [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+      ]]],
+    )
+      .send({
+        from: accounts[0],
+        gas: '1000000',
+      });
+    await adapterRegistry.methods.addTokenAdapters(
+      ['ERC20'],
+      [erc20TokenAdapterAddress, tokenAdapterAddress],
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:88]:
```suggestion
      [erc20TokenAdapterAddress],
```

File: test/YFIAdapter.js [hunk 30]
```diff
@@ -0,0 +1,114 @@
...
+
+  it('should return correct balances', async () => {
+    await adapterRegistry.methods['getBalances(address)'](testAddress)
+      .call()
+      .then((result) => {
+        displayToken(result[0].adapterBalances[0].balances[0].base);
+        assert.deepEqual(result[0].adapterBalances[0].balances[0].base.metadata, yfi);
+        assert.equal(result[0].adapterBalances[0].balances[0].underlying.length, 0);
+        displayToken(result[0].adapterBalances[0].balances[1].base);
+        assert.deepEqual(result[0].adapterBalances[0].balances[1].base.metadata, curve_Y);
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:104]:
```suggestion
        assert.deepEqual(result[0].adapterBalances[0].balances[1].base.metadata, yCrv);
```

File: test/YFIAdapter.js [hunk 31]
```diff
@@ -0,0 +1,114 @@
...
+        assert.equal(result[0].adapterBalances[0].balances[1].underlying.length, 0);
+        displayToken(result[0].adapterBalances[0].balances[2].base);
+        assert.deepEqual(result[0].adapterBalances[0].balances[2].base.metadata, balancer_DAI_YFI_98_2);
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:107]:
```suggestion
        assert.deepEqual(result[0].adapterBalances[0].balances[2].base.metadata, balancerDai98Yfi2);
```

File: test/YFIAdapter.js [hunk 32]
```diff
@@ -0,0 +1,114 @@
...
+        assert.equal(result[0].adapterBalances[0].balances[2].underlying.length, 0);
+        displayToken(result[0].adapterBalances[0].balances[3].base);
+        assert.deepEqual(result[0].adapterBalances[0].balances[3].base.metadata, balancer_YFI_CURVE_Y_2_98);
```
Comment (sobolev-igor) [inline test/YFIAdapter.js:110]:
```suggestion
        assert.deepEqual(result[0].adapterBalances[0].balances[3].base.metadata, balancerYfi2yCrv98);
```

File: contracts/adapters/yearn/YFIAdapter.sol [hunk 33]
```diff
@@ -0,0 +1,81 @@
...
+    function earned(address) external view returns (uint256);
+}
+
+
+/**
+ * @title Adapter for Yearn.finance YFI Staking protocol.
+ * @dev Implementation of ProtocolAdapter interface.
+ * @author Connor Martin <cnr.mrtn@gmail.com>
+ */
+
```
Comment (sobolev-igor) [inline contracts/adapters/yearn/YFIAdapter.sol:39]: remove

File: migrations_scripts/1_deploy_registry_and_add_adapters.js (hunk 10 above)
Comment (sobolev-igor) [reply migrations_scripts/1_deploy_registry_and_add_adapters.js]: lowercase is ok, I changed names a bit in order not to use underscores in names)

File: contracts/adapters/yearn/YFIAdapter.sol [hunk 34]
```diff
@@ -0,0 +1,78 @@
...
+    /**
+     * @return Amount of YFI rewards earned after staking in a given account.
+     * @dev Implementation of ProtocolAdapter interface function.
+     */
+    function getBalance(address token, address account) external view override returns (uint256) {
+      if (token == YFI) {
+          uint256 totalRewards = 0;
+          totalRewards += ERC20(CURVE).earned(account);
+          totalRewards += ERC20(BALANCER).earned(account);
+          totalRewards += ERC20(GOVERNANCE).earned(account);
```
Comment (sobolev-igor) [inline contracts/adapters/yearn/YFIAdapter.sol:64]:
```suggestion
          totalRewards += StakingRewards(CURVE).earned(account);
          totalRewards += StakingRewards(BALANCER).earned(account);
          totalRewards += StakingRewards(GOVERNANCE).earned(account);
```

File: contracts/adapters/yearn/YFIAdapter.sol [hunk 35]
```diff
@@ -0,0 +1,78 @@
...
+          totalRewards += ERC20(FEE_REWARDS).balanceOf(account);
+          return totalRewards;
+      } else if (token == CURVE_Y) {
+          return ERC20(CURVE).balanceOf(account);
+          return StakingRewards(FEE_REWARDS).earned(account);
```
Comment (sobolev-igor) [inline contracts/adapters/yearn/YFIAdapter.sol:69]:
```suggestion
          uint256 totalRewards = 0;
          totalRewards += ERC20(CURVE).balanceOf(account);
          totalRewards += StakingRewards(FEE_REWARDS).earned(account);
          return totalRewards;
```

**Pull request merged**